### PR TITLE
Minimize net diff of binder-gen visibility/identifier-clash fix

### PR DESCRIPTION
**Pull request by layomia**
Follow up to https://github.com/dotnet/runtime/pull/91657. Primarily reverts JSON changes made in that PR.

https://github.com/dotnet/runtime/pull/91967 shows the overall diff of the bug fix and addresses feedback to switch to use the Roslyn API for accessibility checks.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 1]
```diff
@@ -1592,6 +1590,37 @@ private static string DeterminePropertyNameFieldName(string effectiveJsonPropert
...
+                }
+
+                if (type is not INamedTypeSymbol namedType || !namedType.IsGenericType)
+                {
+                    return type.Name;
+                }
+
+                StringBuilder sb = new();
+
+                string name = namedType.Name;
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1609]: >string name = namedType.Name; [](http://example.com/codeflow?start=16&length=29)

why you need to declare this? can't you just append `namedType.Name` to the string builder.

**Review by tarekgh — APPROVED** [review]
LGTM. will be good to wait @eiriktsarpalis to look at the change in the json source gen.

**Comment by layomia** [comment]
Following offline discussion with @eiriktsarpalis, I've addressed his last concern regarding using the built-in API for accessibility checks. https://github.com/dotnet/runtime/pull/91974/commits/3b828344ffbb1147538ba3d178cbeea16e45c405

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/KnownTypeSymbols.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
 {
     internal sealed record KnownTypeSymbols
     {
+        public CSharpCompilation Compilation { get; }
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/KnownTypeSymbols.cs:16]: (This diff has no bearing on generator incrementality.)

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Pull request merged**